test/unit/org/apache/cassandra/utils/memory/NativeAllocatorTest.java

```diff
@@ -22,105 +22,140 @@
+
+    private void verifyUsedReclaiming(long used, long reclaiming)
+    {
+        Assert.assertEquals(used, allocator.offHeap().owns());
+        Assert.assertEquals(used, pool.offHeap.used());
+        Assert.assertEquals(reclaiming, allocator.offHeap().getReclaiming());
+        Assert.assertEquals(reclaiming, pool.offHeap.getReclaiming());
+    }
 
     @Test
     public void testBookKeeping() throws ExecutionException, InterruptedException
     {
-        {
-            final ScheduledExecutorService exec = Executors.newScheduledThreadPool(2);
-            final OpOrder order = new OpOrder();
-            final OpOrder.Group group = order.start();
-            final CountDownLatch canClean = new CountDownLatch(1);
-            final CountDownLatch isClean = new CountDownLatch(1);
-            final AtomicReference<NativeAllocator> allocatorRef = new AtomicReference<>();
-            final AtomicReference<OpOrder.Barrier> barrier = new AtomicReference<>();
-            final NativeAllocator allocator = new NativeAllocator(new NativePool(1, 100, 0.75f, new Runnable()
-            {
-                public void run()
+        final Runnable test = () -> {
+            Assert.assertTrue(allocator.isLive());
+
+            // allocate normal, check accounted and not cleaned
+            allocator.allocate(10, group);
+            verifyUsedReclaiming(10, 0);
+
+            // confirm adjustment works
+            allocator.offHeap().adjust(-10, group);
+            verifyUsedReclaiming(0, 0);
+
+            allocator.offHeap().adjust(10, group);
+            verifyUsedReclaiming(10, 0);
+
+            // confirm we cannot allocate negative
+            boolean success = false;
+            try
                 {
-                    try
-                    {
-                        canClean.await();
-                    }
-                    catch (InterruptedException e)
-                    {
-                        throw new AssertionError();
-                    }
-                    if (isClean.getCount() > 0)
-                    {
-                        allocatorRef.get().offHeap().released(80);
-                        isClean.countDown();
-                    }
+                    allocator.offHeap().allocate(-10, group);
                 }
-            }));
-            allocatorRef.set(allocator);
-            final Runnable markBlocking = new Runnable()
+            catch (AssertionError e)
             {
+                success = true;
+            }
+
+            Assert.assertTrue(success);
+            Uninterruptibles.sleepUninterruptibly(10L, TimeUnit.MILLISECONDS);
+            Assert.assertEquals(1, isClean.getCount());
 
-                public void run()
+            // allocate above watermark
+            allocator.allocate(70, group);
+            verifyUsedReclaiming(80, 0);
+
+            // let the cleaner run, it will release 80 bytes
+            canClean.countDown();
+            try
                 {
```

Comment:
Nit: everything from here to the end of the method is misaligned with an extra tab.